diff --unified --recursive --no-dereference --text --unidirectional-new-file --ignore-trailing-space --ignore-space-change '--exclude=.git' '--exclude=README*' '--exclude=*.pdf' '--exclude=*.svg' '--exclude=*.vim' module_src/iwlwifi_orig/iwl-drv.c module_src/iwlwifi/iwl-drv.c
--- module_src/iwlwifi_orig/iwl-drv.c
+++ module_src/iwlwifi/iwl-drv.c
@@ -1538,6 +1538,13 @@
 		api_ver = drv->fw.ucode_ver;
 	else
 		api_ver = IWL_UCODE_API(drv->fw.ucode_ver);
+ 	/*
+	 * Dan -- disable firmware version checks. Instead, verify
+	 * that the firmware has the size expected for the hacked firmware
+	 */
+	(void)api_min; /* Dan -- remove warning */
+	(void)api_max; /* Fab -- remove warning */
+#if 0
 
 	/*
 	 * api_ver should match the api version forming part of the
@@ -1552,6 +1559,13 @@
 		goto try_again;
 	}
 
+#endif
+
+	if (ucode_raw->size != 336252) {
+		IWL_ERR(drv,
+			"Firmware size does not match iwlwifi-5000-2.ucode.sigcomm2010. The UW 802.11n CSI Tool will not work.\n");
+	}
+
 	/*
 	 * In mvm uCode there is no difference between data and instructions
 	 * sections.
@@ -1880,8 +1894,12 @@
 	.fw_restart = true,
 	.bt_coex_active = true,
 	.power_level = IWL_POWER_INDEX_1,
+#ifdef CONFIG_IWLWIFI_DEBUG
+	.debug_level = IWL_DL_FW_ERRORS,
+#endif
 	.uapsd_disable = IWL_DISABLE_UAPSD_BSS | IWL_DISABLE_UAPSD_P2P_CLIENT,
 	.enable_ini = ENABLE_INI,
+	.connector_log = 0x1,
 	/* the rest are 0 by default */
 };
 IWL_EXPORT_SYMBOL(iwlwifi_mod_params);
@@ -1975,6 +1993,9 @@
 MODULE_PARM_DESC(debug, "debug output mask");
 #endif
 
+module_param_named(connector_log, iwlwifi_mod_params.connector_log, int, S_IRUGO);
+MODULE_PARM_DESC(connector_log, "set connector log mask (default 1 [BFEE capture])");
+
 module_param_named(swcrypto, iwlwifi_mod_params.swcrypto, int, 0444);
 MODULE_PARM_DESC(swcrypto, "using crypto in software (default 0 [hardware])");
 module_param_named(11n_disable, iwlwifi_mod_params.disable_11n, uint, 0444);
