diff --unified --recursive --no-dereference --text --unidirectional-new-file --ignore-trailing-space --ignore-space-change '--exclude=.git' '--exclude=README*' '--exclude=*.pdf' '--exclude=*.svg' '--exclude=*.vim' module_src/iwlwifi_orig/dvm/dev.h module_src/iwlwifi/dvm/dev.h
--- module_src/iwlwifi_orig/dvm/dev.h
+++ module_src/iwlwifi/dvm/dev.h
@@ -258,7 +258,7 @@
 /* Sensitivity and chain noise calibration */
 #define INITIALIZATION_VALUE		0xFFFF
 #define IWL_CAL_NUM_BEACONS		16
-#define MAXIMUM_ALLOWED_PATHLOSS	15
+#define MAXIMUM_ALLOWED_PATHLOSS	50
 
 #define CHAIN_NOISE_MAX_DELTA_GAIN_CODE 3
 
@@ -787,6 +787,7 @@
 	u8 agg_tids_count;
 
 	struct iwl_rx_phy_res last_phy_res;
+	u8 last_cfg_phy_buf[IWLAGN_MAX_CFG_PHY_CNT*sizeof(u32)];
 	u32 ampdu_ref;
 	bool last_phy_res_valid;
 
@@ -852,6 +853,21 @@
 	void *wowlan_sram;
 #endif /* CONFIG_IWLWIFI_DEBUGFS */
 
+	u32 connector_log;
+	int last_rx_noise;
+	s8 last_rx_noiseA;
+	s8 last_rx_noiseB;
+	s8 last_rx_noiseC;
+	bool bf_enabled;
+	u32 monitor_tx_rate;
+	u32 bcast_tx_rate;
+
+	/* Setup for rotating rates */
+	u32 rotate_rates;
+	u32 last_rotate_rate;
+	u32 rotate_rate_total;
+	u32 *rotate_rate_array;
+
 	struct iwl_nvm_data *nvm_data;
 	/* eeprom blob for debugfs */
 	u8 *eeprom_blob;
@@ -888,6 +904,9 @@
 
 	/* indicator of loaded ucode image */
 	enum iwl_ucode_type cur_ucode;
+
+	/* Remember last encountered sequence number */
+	u16 last_seq_nr;
 }; /*iwl_priv */
 
 static inline struct iwl_rxon_context *
