diff --unified --recursive --no-dereference --text --unidirectional-new-file --ignore-trailing-space --ignore-space-change '--exclude=.git' '--exclude=README*' '--exclude=*.pdf' '--exclude=*.svg' '--exclude=*.vim' module_src/iwlwifi_orig/dvm/main.c module_src/iwlwifi/dvm/main.c
--- module_src/iwlwifi_orig/dvm/main.c
+++ module_src/iwlwifi/dvm/main.c
@@ -37,7 +37,7 @@
 #include "dev.h"
 #include "calib.h"
 #include "agn.h"
-
+#include "connector.h"
 
 /******************************************************************************
  *
@@ -619,7 +619,7 @@
 	priv->contexts[IWL_RXON_CTX_BSS].exclusive_interface_modes =
 		BIT(NL80211_IFTYPE_ADHOC) | BIT(NL80211_IFTYPE_MONITOR);
 	priv->contexts[IWL_RXON_CTX_BSS].interface_modes =
-		BIT(NL80211_IFTYPE_STATION);
+		BIT(NL80211_IFTYPE_STATION) | BIT(NL80211_IFTYPE_AP);
 	priv->contexts[IWL_RXON_CTX_BSS].ap_devtype = RXON_DEV_TYPE_AP;
 	priv->contexts[IWL_RXON_CTX_BSS].ibss_devtype = RXON_DEV_TYPE_IBSS;
 	priv->contexts[IWL_RXON_CTX_BSS].station_devtype = RXON_DEV_TYPE_ESS;
@@ -853,6 +853,8 @@
 	/* At this point, the NIC is initialized and operational */
 	iwl_rf_kill_ct_config(priv);
 
+	iwl_connector_set_priv(priv);
+
 	IWL_DEBUG_INFO(priv, "ALIVE processing complete.\n");
 
 	return iwl_power_update_mode(priv, true);
@@ -1103,6 +1105,16 @@
 	priv->missed_beacon_threshold = IWL_MISSED_BEACON_THRESHOLD_DEF;
 	priv->agg_tids_count = 0;
 
+	/* Dan's parameters */
+	priv->connector_log = iwlwifi_mod_params.connector_log;
+	priv->bf_enabled = 1;			/* Enabled */
+	priv->rotate_rates = 0;			/* Disabled */
+	priv->last_rotate_rate = 0;		/* Disabled */
+	priv->rotate_rate_total = 0;	        /* Disabled */
+	priv->rotate_rate_array = NULL;	        /* Disabled */
+	priv->monitor_tx_rate = 0;		/* Disabled */
+	priv->bcast_tx_rate = 0;		/* Disabled */
+
 	priv->rx_statistics_jiffies = jiffies;
 
 	/* Choose which receivers/antennas to use */
@@ -1240,6 +1252,7 @@
 		REPLY_COMPRESSED_BA,
 		STATISTICS_NOTIFICATION,
 		REPLY_TX,
+		REPLY_BFEE_NOTIFICATION,
 	};
 	int i;
 
@@ -2140,10 +2153,19 @@
 		return ret;
 	}
 
+	ret = iwlagn_register_connector();
+	if (ret) {
+		pr_err("Unable to initialize connector: %d\n", ret);
+		iwlagn_rate_control_unregister();
+		return ret;
+	}
+
 	ret = iwl_opmode_register("iwldvm", &iwl_dvm_ops);
 	if (ret) {
 		pr_err("Unable to register op_mode: %d\n", ret);
+		iwlagn_unregister_connector();
 		iwlagn_rate_control_unregister();
+		return ret;
 	}
 
 	return ret;
@@ -2153,6 +2175,7 @@
 static void __exit iwl_exit(void)
 {
 	iwl_opmode_deregister("iwldvm");
+	iwlagn_unregister_connector();
 	iwlagn_rate_control_unregister();
 }
 module_exit(iwl_exit);
