diff --unified --recursive --no-dereference --text --unidirectional-new-file --ignore-trailing-space --ignore-space-change '--exclude=.git' '--exclude=README*' '--exclude=*.pdf' '--exclude=*.svg' '--exclude=*.vim' module_src/ath_orig/ath9k/Makefile module_src/ath/ath9k/Makefile
--- module_src/ath_orig/ath9k/Makefile
+++ module_src/ath/ath9k/Makefile
@@ -1,4 +1,5 @@
 # SPDX-License-Identifier: ISC
+obj-m   += 	ar9003_csi.o 
 ath9k-y +=	beacon.o \
 		gpio.o \
 		init.o \
