diff --unified --recursive --no-dereference --text --unidirectional-new-file --ignore-trailing-space --ignore-space-change '--exclude=.git' '--exclude=README*' '--exclude=*.pdf' '--exclude=*.svg' '--exclude=*.vim' module_src/iwlwifi_original/dvm/Makefile module_src/iwlwifi/dvm/Makefile
--- module_src/iwlwifi_original/dvm/Makefile	2023-04-13 16:55:55.783843265 +0200
+++ module_src/iwlwifi/dvm/Makefile	2023-03-16 20:02:29.025683167 +0100
@@ -4,7 +4,7 @@
 iwldvm-objs		+= lib.o calib.o tt.o sta.o rx.o
 
 iwldvm-objs		+= power.o
-iwldvm-objs		+= scan.o
+iwldvm-objs		+= scan.o connector.o
 iwldvm-objs		+= rxon.o devices.o
 
 iwldvm-$(CONFIG_IWLWIFI_LEDS) += led.o
