diff --unified --recursive --no-dereference --text --unidirectional-new-file --ignore-trailing-space --ignore-space-change '--exclude=.git' '--exclude=README*' '--exclude=*.pdf' '--exclude=*.svg' '--exclude=*.vim' module_src/iwlwifi_original/dvm/rx.c module_src/iwlwifi/dvm/rx.c
--- module_src/iwlwifi_original/dvm/rx.c	2023-04-13 16:55:55.790509932 +0200
+++ module_src/iwlwifi/dvm/rx.c	2023-03-16 20:02:29.029016500 +0100
@@ -36,6 +36,7 @@
 #include "dev.h"
 #include "calib.h"
 #include "agn.h"
+#include "connector.h"
 
 #define IWL_CMD_ENTRY(x) [x] = #x
 
@@ -114,6 +115,7 @@
 	IWL_CMD_ENTRY(REPLY_WOWLAN_KEK_KCK_MATERIAL),
 	IWL_CMD_ENTRY(REPLY_WOWLAN_GET_STATUS),
 	IWL_CMD_ENTRY(REPLY_D3_CONFIG),
+	IWL_CMD_ENTRY(REPLY_BFEE_NOTIFICATION),
 };
 #undef IWL_CMD_ENTRY
 
@@ -372,10 +374,17 @@
 	}
 
 	/* Average among active antennas */
-	if (num_active_rx)
+	if (num_active_rx) {
 		last_rx_noise = (total_silence / num_active_rx) - 107;
-	else
+		if (priv->connector_log & IWL_CONN_NOISE_MSK)
+			connector_send_msg((void *)&last_rx_noise,
+					sizeof(last_rx_noise), IWL_CONN_NOISE);
+	} else
 		last_rx_noise = IWL_NOISE_MEAS_NOT_AVAILABLE;
+	priv->last_rx_noise = last_rx_noise;
+	priv->last_rx_noiseA = bcn_silence_a;
+	priv->last_rx_noiseB = bcn_silence_b;
+	priv->last_rx_noiseC = bcn_silence_c;
 
 	IWL_DEBUG_CALIB(priv, "inband silence a %u, b %u, c %u, dBm %d\n",
 			bcn_silence_a, bcn_silence_b, bcn_silence_c,
@@ -659,11 +668,14 @@
 				struct iwl_device_cmd *cmd)
 {
 	struct iwl_rx_packet *pkt = rxb_addr(rxb);
+	struct iwl_rx_phy_res *rx_phy_res = (void *)pkt->data;
 
 	priv->last_phy_res_valid = true;
 	priv->ampdu_ref++;
 	memcpy(&priv->last_phy_res, pkt->data,
 	       sizeof(struct iwl_rx_phy_res));
+	memcpy(&priv->last_cfg_phy_buf, rx_phy_res->cfg_phy_buf,
+			rx_phy_res->cfg_phy_cnt * sizeof(u32));
 	return 0;
 }
 
@@ -915,10 +927,12 @@
 	rx_pkt_status = *(__le32 *)(pkt->data + sizeof(*amsdu) + len);
 	ampdu_status = iwlagn_translate_rx_status(priv,
 						  le32_to_cpu(rx_pkt_status));
+	if (priv->connector_log & IWL_CONN_RX_MPDU_MSK)
+		connector_send_msg((void *)header, len, IWL_CONN_RX_MPDU);
 
-	if ((unlikely(phy_res->cfg_phy_cnt > 20))) {
-		IWL_DEBUG_DROP(priv, "dsp size out of range [0,20]: %d\n",
-				phy_res->cfg_phy_cnt);
+	if ((unlikely(phy_res->cfg_phy_cnt > IWLAGN_MAX_CFG_PHY_CNT))) {
+		IWL_DEBUG_DROP(priv, "dsp size out of range [0,%d]: %d\n",
+				IWLAGN_MAX_CFG_PHY_CNT, phy_res->cfg_phy_cnt);
 		return 0;
 	}
 
@@ -1092,6 +1106,9 @@
 	handlers[REPLY_COMPRESSED_BA]		=
 		iwlagn_rx_reply_compressed_ba;
 
+	/* Beamforming */
+	handlers[REPLY_BFEE_NOTIFICATION] = iwlagn_bfee_notif;
+
 	priv->rx_handlers[REPLY_TX] = iwlagn_rx_reply_tx;
 
 	/* set up notification wait support */
